File: EyePal/ViewModels/LyricPrompterViewModel.swift
import Foundation
import UIKit

@MainActor
final class LyricPrompterViewModel: ObservableObject {
    @Published var searchText = ""
    @Published var savedSongs: [LyricSong] = []
    @Published var currentSong: LyricSong?
    @Published var searchResults: [LyricSearchResult] = []
    @Published var searchLog: [String] = []
    @Published var isSearching = false
    @Published var isLoadingLyrics = false
    @Published var errorMessage: String?
    @Published var isPlaying = false

    private let service = LyricPrompterService()
    private let announcer = AccessibilityAnnouncementCenter()
    private var playbackTask: Task<Void, Never>?
    private var advanceOffset: TimeInterval = 0
    private weak var settingsStore: SettingsStore?
    private weak var openAIStore: OpenAISubscriptionStore?

    var savedSongsURL: URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let base = appSupport.appendingPathComponent("EyePal", isDirectory: true)
        return base.appendingPathComponent("lyrics.json")
    }

    func bind(settings: SettingsStore, openAIStore: OpenAISubscriptionStore) {
        self.settingsStore = settings
        self.openAIStore = openAIStore
    }

    func loadSaved() {
        guard FileManager.default.fileExists(atPath: savedSongsURL.path),
              let data = try? Data(contentsOf: savedSongsURL),
              let songs = try? JSONDecoder().decode([LyricSong].self, from: data) else {
            savedSongs = []
            return
        }
        savedSongs = songs
    }

    func saveCurrentSong() {
        guard let song = currentSong else { return }
        if let index = savedSongs.firstIndex(where: { $0.id == song.id }) {
            savedSongs[index] = song
        } else {
            savedSongs.insert(song, at: 0)
        }
        persistSongs()
    }

    func deleteSong(_ song: LyricSong) {
        savedSongs.removeAll { $0.id == song.id }
        if currentSong?.id == song.id { currentSong = nil }
        persistSongs()
    }

    func selectSong(_ song: LyricSong) {
        currentSong = song
        searchText = "\(song.title) - \(song.artist)"
    }

    func search() async {
        let input = searchText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !input.isEmpty else { return }

        guard let settings = settingsStore else {
            errorMessage = "Settings unavailable."
            return
        }

        isSearching = true
        errorMessage = nil
        searchResults = []
        searchLog = ["Searching: \(input)"]

        let results = await service.searchAllSources(
            searchText: input,
            provider: LyricLLMProvider(rawValue: settings.lyricLLMProvider) ?? .codex,
            modelID: settings.lyricModelID,
            apiKey: settings.lyricAPIKey,
            baseURL: settings.lyricBaseURL,
            codexStore: openAIStore
        )

        if results.isEmpty {
            searchLog.append("No results found.")
            errorMessage = "No results found for \"\(input)\"."
            announcer.announce("No results found for \(input)", minimumInterval: 0)
            isSearching = false
        } else {
            for r in results {
                searchLog.append("[\(r.source)] \(r.trackName) - \(r.artistName) synced=\(r.hasSyncedLyrics)")
            }
            searchResults = results
            isSearching = false
        }
    }

    func loadSelectedResult(_ result: LyricSearchResult) {
        isLoadingLyrics = true
        currentSong = service.loadLyrics(from: result)
        isLoadingLyrics = false
    }

    func dismissResults() {
        searchResults = []
    }

    func playFromStart(offset: TimeInterval) {
        guard let song = currentSong, song.hasTimestamps else { return }
        stopPlayback()
        isPlaying = true
        advanceOffset = offset

        playbackTask = Task { [weak self] in
            guard let self else { return }
            let linesWithTime = song.lines.filter { $0.startTime != nil }
            guard let firstTime = linesWithTime.first?.startTime else {
                await MainActor.run { self.isPlaying = false }
                return
            }
            let waitTime = max(0, firstTime - offset)
            try? await Task.sleep(nanoseconds: UInt64(waitTime * 1_000_000_000))
            guard !Task.isCancelled else { return }
            for line in linesWithTime {
                guard !Task.isCancelled else { break }
                await MainActor.run { self.announcer.announce(line.text, minimumInterval: 0) }
                if let nextTime = self.nextTimestamp(after: line.startTime!, in: linesWithTime) {
                    let delay = max(0, nextTime - line.startTime! - offset)
                    try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                }
            }
            await MainActor.run { self.isPlaying = false }
        }
    }

    func playFromNow(offset: TimeInterval) {
        guard let song = currentSong, song.hasTimestamps else { return }
        stopPlayback()
        isPlaying = true
        advanceOffset = offset

        // Focus will be moved by the view via @AccessibilityFocusState
        // The view observes isPlaying and focuses first line

        playbackTask = Task { [weak self] in
            guard let self else { return }
            let linesWithTime = song.lines.filter { $0.startTime != nil }
            guard linesWithTime.count >= 2 else {
                await MainActor.run { self.isPlaying = false }
                return
            }
            // First line is already focused by the view — VoiceOver reads it.
            // Start the timer from the first line's timestamp so we announce
            // the second line when its time arrives.
            for i in 1..<linesWithTime.count {
                guard !Task.isCancelled else { break }
                let prev = linesWithTime[i - 1]
                let curr = linesWithTime[i]
                let delay = max(0, curr.startTime! - prev.startTime! - offset)
                try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                guard !Task.isCancelled else { break }
                await MainActor.run { self.announcer.announce(curr.text, minimumInterval: 0) }
            }
            await MainActor.run { self.isPlaying = false }
        }
    }

    func stopPlayback() {
        playbackTask?.cancel()
        playbackTask = nil
        isPlaying = false
    }

    private func nextTimestamp(after time: Double, in lines: [LyricLine]) -> Double? {
        lines.first { $0.startTime! > time }?.startTime
    }

    private func persistSongs() {
        guard let data = try? JSONEncoder.prettyPrinted.encode(savedSongs) else { return }
        try? data.write(to: savedSongsURL, options: .atomic)
    }
}

private extension JSONEncoder {
    static var prettyPrinted: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }
}

File: EyePalAndroid/app/src/main/java/com/eyepal/app/viewmodels/LyricPrompterViewModel.kt
package com.eyepal.app.viewmodels

import android.app.Application
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel
import kotlinx.coroutines.flow.first
import androidx.lifecycle.viewModelScope
import com.eyepal.app.data.SettingsRepository
import com.eyepal.app.models.LyricSong
import com.eyepal.app.services.AccessibilityAnnouncer
import com.eyepal.app.services.LyricPrompterService
import com.eyepal.app.services.LyricStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class LyricPrompterViewModel(application: Application) : AndroidViewModel(application) {
    val searchText = mutableStateOf("")
    val searchResults = mutableStateOf<List<com.eyepal.app.models.LyricSearchResult>>(emptyList())
    val currentSong = mutableStateOf<LyricSong?>(null)
    val isSearching = mutableStateOf(false)
    val isLoadingLyrics = mutableStateOf(false)
    val isPlaying = mutableStateOf(false)
    val currentLineIndex = mutableStateOf(-1)
    val errorMessage = mutableStateOf<String?>(null)
    val savedSongs = mutableStateOf<List<LyricSong>>(emptyList())

    private val service = LyricPrompterService(context = application)
    private val settings = SettingsRepository(application)
    private val announcer = AccessibilityAnnouncer(application)
    private var playbackJob: Job? = null

    init {
        savedSongs.value = LyricStorage.loadSongs(application)
    }

    fun search() {
        val input = searchText.value.trim()
        if (input.isEmpty()) return
        isSearching.value = true
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val results = service.searchAllSources(input)
                if (results.isEmpty()) { errorMessage.value = "No lyrics found." }
                else { searchResults.value = results }
            } catch (e: Exception) { errorMessage.value = e.message }
            isSearching.value = false
        }
    }

    fun loadSelectedResult(result: com.eyepal.app.models.LyricSearchResult) {
        isLoadingLyrics.value = true
        currentSong.value = service.loadLyrics(result)
        isLoadingLyrics.value = false
    }

    fun loadSavedSong(song: LyricSong) {
        currentSong.value = song
    }

    fun loadSavedSongs() {
        savedSongs.value = LyricStorage.loadSongs(getApplication())
    }

    fun saveCurrentSong() {
        val song = currentSong.value ?: return
        val savedSong = song.copy(id = System.currentTimeMillis().toString())
        savedSongs.value = listOf(savedSong) + savedSongs.value
        LyricStorage.addSong(getApplication(), savedSong)
        announcer.announce("${song.title} by ${song.artist} saved.")
    }

    fun deleteSong(songId: String) {
        savedSongs.value = savedSongs.value.filter { it.id != songId }
        LyricStorage.deleteSong(getApplication(), songId)
    }

    fun playFromStart() {
        val song = currentSong.value ?: return
        if (!song.hasTimestamps) return
        playFromLine(0)
    }

    fun playFromNow() {
        val song = currentSong.value ?: return
        if (!song.hasTimestamps) return
        stopPlayback()
        isPlaying.value = true
        playbackJob = viewModelScope.launch {
            val linesWithTime = song.lines.filter { it.startTime != null }
            val advanceOffset = settings.lyricAdvanceOffset.first()
            if (linesWithTime.isEmpty()) { isPlaying.value = false; return@launch }
            currentLineIndex.value = song.lines.indexOf(linesWithTime[0])
            announcer.announce(linesWithTime[0].text)
            for (i in 1 until linesWithTime.size) {
                if (!isPlaying.value) break
                val prev = linesWithTime[i - 1]
                val curr = linesWithTime[i]
                delay(((curr.startTime!! - prev.startTime!! - advanceOffset).coerceAtLeast(0.0) * 1000).toLong())
                currentLineIndex.value = song.lines.indexOf(curr)
                announcer.announce(curr.text)
            }
            currentLineIndex.value = -1
            isPlaying.value = false
        }
    }

    fun playFromLine(index: Int) {
        val song = currentSong.value ?: return
        if (!song.hasTimestamps) return
        stopPlayback()
        isPlaying.value = true
        playbackJob = viewModelScope.launch {
            val linesWithTime = song.lines.filter { it.startTime != null }
            val advanceOffset = settings.lyricAdvanceOffset.first()
            if (linesWithTime.isEmpty()) { isPlaying.value = false; return@launch }

            // Find the timed line closest to or at the given index
            val targetLine = linesWithTime.firstOrNull { song.lines.indexOf(it) >= index }
                ?: linesWithTime.lastOrNull()
                ?: return@launch
            val startIndex = linesWithTime.indexOf(targetLine)

            currentLineIndex.value = song.lines.indexOf(targetLine)
            announcer.announce(targetLine.text)

            for (i in (startIndex + 1) until linesWithTime.size) {
                if (!isPlaying.value) break
                val prev = linesWithTime[i - 1]
                val curr = linesWithTime[i]
                delay(((curr.startTime!! - prev.startTime!! - advanceOffset).coerceAtLeast(0.0) * 1000).toLong())
                currentLineIndex.value = song.lines.indexOf(curr)
                announcer.announce(curr.text)
            }
            currentLineIndex.value = -1
            isPlaying.value = false
        }
    }

    fun stopPlayback() {
        playbackJob?.cancel()
        playbackJob = null
        isPlaying.value = false
        currentLineIndex.value = -1
    }
}
